### PR TITLE
Approval program pages

### PR DESCRIPTION
**Pull request by sczembor**
Closes #x

## Proposed Changes

- Added support for new txn opcode fileds `ApprovalProgramPages`, `ClearProgramStatePages`, `NumApprovalProgramPages`, `NumClearProgramStatePages`
- Added additional checks for the max `stackByteElementSize` and `maxProgramLength`
- Test scenario for all the features mention above.

## TODO

- [x] update change log

### CODE REVIEW

File: packages/runtime/src/errors/errors-list.ts [hunk 1]
```diff
@@ -439,8 +441,21 @@ maximun uint128`,
...
 		title: "Unknown key JSON",
-		description: "Unknown key JSON"
-	}
+		description: "Unknown key JSON",
+	},
+	MAX_BYTE_ARRAY_EXCEEDED: {
+		number: 1065,
+		message: "Maximum byte array length exceeded",
+		title: "Maximum byte array length exceeded",
+		description: "Maximum byte array length is 4096",
```
Comment (rmeena840) [inline packages/runtime/src/errors/errors-list.ts:450]: Is it possible to use the constant that you have defined for 4096 here?

File: packages/runtime/src/errors/errors-list.ts [hunk 2]
```diff
@@ -439,8 +441,21 @@ maximun uint128`,
...
+	},
+	PROGRAM_LENGTH_EXCEEDED: {
+		number: 1066,
+		message: "Maximum program length exceeded",
+		title: "Program length",
+		description:
+			"Maximum program length (2048 bytes) exceeded. Use `ApprovalProgramPages` instead",
```
Comment (rmeena840) [inline packages/runtime/src/errors/errors-list.ts:457]: Same here for 2048.

File: packages/runtime/src/lib/itxn.ts [hunk 3]
```diff
@@ -283,6 +288,32 @@ export function setInnerTxField(
...
+			if (approvalProgram !== undefined) {
+				//append
+				txValue = new Uint8Array([...approvalProgram, ...(txValue as Uint8Array)]);
+			}
+			if ((txValue as Uint8Array).length > maxPossible) {
+				errMsg = `Approval Program exceeded the maximum allowed length of ${maxPossible} bytes`;
+			}
+			break;
+		}
+		case "ClearStateProgramPages": {
```
Comment (rmeena840) [inline packages/runtime/src/lib/itxn.ts:304]: same here

File: packages/runtime/src/lib/txn.ts [hunk 4]
```diff
@@ -54,23 +55,38 @@ const globalAndLocalNumTxnFields = new Set([
...
-
 	if (typeof a === "string") {
-		return parsing.stringToBytes(a);
+		byteArray = parsing.stringToBytes(a);
+	}
+	if (byteArray.length > maxByteArrayLength) {
+		throw new RuntimeError(RUNTIME_ERRORS.TEAL.MAX_BYTE_ARRAY_EXCEEDED);
+	}
+	if (
+		(field === "ApprovalProgram" || field === "ClearStateProgram") &&
```
Comment (rmeena840) [inline packages/runtime/src/lib/txn.ts:84]: same here. Better create enum for this.

File: packages/runtime/src/parser/parser.ts [hunk 5]
```diff
@@ -860,6 +860,8 @@ function _assertMaxLen(len: number, mode: ExecutionMode): void {
 		}
 	} else {
 		if (len > MaxAppProgramLen) {
+			// TODO: // When MaxExtraAppProgramPages > 0, this is the size of those pages.
```
Comment (rmeena840) [inline packages/runtime/src/parser/parser.ts:863]: make sure we created a task for this in pivotal

**Review by rmeena840 — COMMENTED** [review]
Reviewed

File: packages/runtime/src/lib/constants.ts [hunk 6]
```diff
@@ -10,6 +10,7 @@ export const MAX_UINT6 = 63n;
 export const DEFAULT_STACK_ELEM = 0n;
 export const MAX_CONCAT_SIZE = 4096;
 export const ALGORAND_MIN_TX_FEE = 1000;
+export const maxByteArrayLength = 4096;
```
Comment (vuvoth) [inline packages/runtime/src/lib/constants.ts:13]: Let use right name convention.

File: packages/runtime/src/lib/txn.ts [hunk 7]
```diff
@@ -225,6 +252,36 @@ export function txAppArg(
...
+	} else if (txField === "ApprovalProgramPages" && tx.apap) {
+		const pageCount = Math.ceil(tx.apap.length / maxByteArrayLength);
+		if (idx > pageCount) {
+			throw new Error("invalid ApprovalProgramPages page index");
+		}
+		const first = idx * maxByteArrayLength;
+		let last = first + maxByteArrayLength;
+		if (last > tx.apap.length) {
+			last = tx.apap.length;
+		}
```
[the remark this answers is not in the document]
Comment (vuvoth) [reply packages/runtime/src/lib/txn.ts:264]: Or maybe `const last = min(first + maxByteArrayLength, tx.apap.length)`

File: packages/runtime/src/lib/txn.ts [hunk 8]
```diff
@@ -225,6 +252,36 @@ export function txAppArg(
...
+	} else if (txField === "NumApprovalProgramPages" && tx.apap) {
+		const pageCount = Math.ceil(tx.apap.length / maxByteArrayLength);
+		return parseToStackElem(pageCount, txField);
+	} else if (txField === "ClearStateProgramPages" && tx.apsu) {
+		const pageCount = Math.ceil(tx.apsu.length / maxByteArrayLength);
+		if (idx > pageCount) {
+			throw new Error("invalid ClearStateProgramPages page index");
+		}
+		const first = idx * maxByteArrayLength;
+		let last = first + maxByteArrayLength;
```
Comment (vuvoth) [inline packages/runtime/src/lib/txn.ts:276]: Same here!

**Review by vuvoth — COMMENTED** [review]
Thanks! I have a few comments!

File: packages/runtime/src/lib/constants.ts (hunk 6 above)
Comment (sczembor) [reply packages/runtime/src/lib/constants.ts:13]: In the goAlgorand the use `maxStringSize`

**Review by rmeena840 — APPROVED** [review]

File: packages/runtime/src/lib/txn.ts [hunk 9]
```diff
@@ -119,78 +142,87 @@ export function txnSpecByField(
...
-		case "StateProofPK": {
+		case TxFieldEnum.StateProofPK: {
 			// While running teal debugger, "StateProofPK" always return 64 zero bytes.
 			// so we set up 64 zero bytes as default value of StateProofPK
 			result = new Uint8Array(64).fill(0); // 64 zero bytes
 			break;
 		}
-
+		case TxFieldEnum.NumApprovalProgramPages: {
+			if (tx.apap) result = Math.ceil(tx.apap.length / maxStringSize);
```
Comment (rmeena840) [inline packages/runtime/src/lib/txn.ts:217]: use ternary operator here

File: packages/runtime/src/lib/txn.ts [hunk 10]
```diff
@@ -119,78 +142,87 @@ export function txnSpecByField(
...
+			else result = 0n;
+			break;
+		}
+		case TxFieldEnum.NumClearStateProgramPages: {
+			if (tx.apsu) result = Math.ceil(tx.apsu.length / maxStringSize);
```
Comment (rmeena840) [inline packages/runtime/src/lib/txn.ts:222]: same here

**Review by rmeena840 — COMMENTED** [review]
check comment before merge

**Comment by rmeena840** [comment]
@sczembor Update changelog.md before merge. THanks

**Pull request merged**